include/dca/math/nfft/dnfft_1d_gpu.hpp

```diff
@@ -0,0 +1,245 @@
+  Dnfft1DGpu(double beta, cudaStream_t stream, bool accumulate_m_sqr = false);
+
+  // Resets the accumulated quantities. To be called before each DCA iteration.
+  void initialize();
+
+  // Accumulates asynchronously on the device the entries of the M matrix at times and orbitals
+  // described by the provided configuration.
+  // Postcondition: M and config shall not be modified untill 'synchronizeCopy' is called.
+  template <class Configuration>
+  void accumulate(const linalg::Matrix<double, linalg::CPU>& M, const Configuration& config,
+                  const int sign);
+
+  // Transforms the accumulated data to the frequency domain and stores it in f_w.
+  // If get_square is true, transforms the accumulated squared data instead.
+  template <typename OtherScalarType>
+  void finalize(func::function<std::complex<OtherScalarType>, func::dmn_variadic<WDmn, PDmn>>& f_w,
+                bool get_square = false);
+
+  // Sums the accumulated data in the time domain.
+  ThisType& operator+=(ThisType& other);
+
+  // Ensures that the arguments of the method 'accumulate' have been copied to the GPU.
+  void synchronizeCopy() {
+    m_copied_event_.block();
+  }
+
+private:
+  void initializeDeviceCoefficients();
+
+private:
+  using BaseClass::f_tau_;
+  static inline linalg::Vector<ScalarType, linalg::GPU>& get_device_cubic_coeff();
+
+  const double beta_;
+  cudaStream_t stream_;
+  const bool accumulate_m_sqr_;
+  linalg::Matrix<ScalarType, linalg::GPU> accumulation_matrix_;
+  linalg::Matrix<ScalarType, linalg::GPU> accumulation_matrix_sqr_;
+
+  linalg::Matrix<double, linalg::GPU> M_;
+  linalg::util::HostVector<details::ConfigElem> config_left_;
+  linalg::util::HostVector<details::ConfigElem> config_right_;
+  linalg::util::HostVector<ScalarType> times_;
+  linalg::Vector<details::ConfigElem, linalg::GPU> config_left_dev_;
+  linalg::Vector<details::ConfigElem, linalg::GPU> config_right_dev_;
+  linalg::Vector<ScalarType, linalg::GPU> times_dev_;
+
+  linalg::util::CudaEvent config_copied_event_;
+  linalg::util::CudaEvent m_copied_event_;
+};
+
+template <typename ScalarType, typename WDmn, typename RDmn, int oversampling>
+Dnfft1DGpu<ScalarType, WDmn, RDmn, oversampling, CUBIC>::Dnfft1DGpu(const double beta,
+                                                                    cudaStream_t stream,
+                                                                    const bool accumulate_m_sqr)
+    : BaseClass(), beta_(beta), stream_(stream), accumulate_m_sqr_(accumulate_m_sqr) {
+  initializeDeviceCoefficients();
+  assert(cudaPeekAtLastError() == cudaSuccess);
+  func::dmn_variadic<BDmn, BDmn, RDmn> bbr_dmn;
+  const int n_times = BaseClass::PaddedTimeDmn::dmn_size();
+  accumulation_matrix_.resizeNoCopy(std::make_pair(n_times, bbr_dmn.get_size()));
+  if (accumulate_m_sqr)
+    accumulation_matrix_sqr_.resizeNoCopy(accumulation_matrix_.size());
+}
+
+template <typename ScalarType, typename WDmn, typename RDmn, int oversampling>
```

Comment:
From the use of this function, I think it is better renamed as "resetAccumulationMatrix()" or alike.

Reply:
Renamed to resetAccumulation in e7f136b4333ff98dc2b008cbf1f7694c72a4ea3f to keep the same name in the CPU/GPU version. I might as well us a vector or any other container for the accumulated quantities.